### DISCUSSION TITLE
[6 Luglio 2020] Cache

### DISCUSSION THREAD

**Discussion by Elia-Belli** [sapienzastudentsnetwork/architettura-degli-elaboratori#54] [Cache (M-Z)]
![Esame 6](https://github.com/Elia-Belli/ae2223/assets/107944476/b2ee49fe-dab6-4bb3-8d85-85a257205294)
---
### Soluzione del Prof
Le soluzioni fornite dal professore potrebbero essere incomplete o poco chiare, quindi sono benvenute soluzioni che le completano, oppure soluzioni alternative e/o con spiegazioni più approfondite: [Soluzione.pdf](https://github.com/sapienzastudentsnetwork/archelab2223/files/11665793/AE.-.Appello.2020-07-06.-.Cache.soluzione.pdf)

**Comment by FeddyLix17** [comment]
## 1) Accessi in memoria

dopo essersi ricavati le formule usando un approccio grafico

![image](https://github.com/sapienzastudentsnetwork/archelab2223/assets/66442720/f456dac8-b95a-4da2-94a9-a3b5e55034fe)

si riempie la tabella ottenendo

![image](https://github.com/sapienzastudentsnetwork/archelab2223/assets/66442720/c4f27433-d473-45a4-adac-8d194cdd17b1)


## 2) Dimensioni delle cache

considerando che ogni linea della cache è formata

- dal bit di validità

- dal tag

- dal blocco stesso

si ottiene, rispettivamente per le 2 cache
 
- $\Large \text{L1}$
  
   il numero delle linee è dato dal prodotto tra le vie e i set, da cui

$$
    \Large  (1 + 27 + 4 * 4 * 8) * (4 * 2) \implies (28 + 128) * 8 \implies  156 * 8 \implies 1248\ \text{bit}
$$

- $\Large \text{L2}$

$$
    \Large  (1 + 23 + 16 * 4 * 8) * 8 \implies (24 + 512) * 8 \implies  536 * 8 \implies 4288\ \text{bit}
$$

## 3) Tempi della sequenza di accessi

### (a) Tempo totale sequenza di accessi

rielencandoli brevemente dalla foto superiore, si hanno

- 7 accessi in memoria, con tempo $\Large 200 * 7 = 1400\ \text{ns}$

- 4 accessi in L1, con tempo  $\Large 1 * 4 = 4\ \text{ns}$

- 1 accessi in L2, con tempo  $\Large 20  \text{ns}$

$$
    \Large T_{tot} = 1400 + 4 + 20 = 1424\ \text{ns}
$$

### (b) Tempo medio sequenza di accessi

$$
    \Large T_{med} = \frac{\text{T.\ Totale}}{\text{Num.\ accessi}} = \frac{1424}{12} = 118.67\ \text{ns}
$$

### (c) Istruzioni svolte nel tempo medio

$$
    \Large \text{Istruzioni\ eseguite}  = \frac{\text{T.\ medio} *\ \text{Frequenza}}{\text{Clock\ per\ Istruction}} = \frac{118.67 * 2}{4} = 59.33\
$$

**Reply by matypist** [reply]
I risultati della soluzione proposta da @FeddyLix17 coincidono con i miei e con quelli della soluzione del professore a meno del calcolo della **dimensione** della **cache L1** (**punto n. 2** dell'esercizio), dove **probabilmente bisognerebbe considerare anche** i **bit di controllo** (uno per blocco, quindi $8$) per la **LRU**, ottenendo così una dimensione di $1256$ bit anziché di $1248$ bit.

**Comment by M1SC0** [comment]
![tempFileForShare_20250602-173245](https://github.com/user-attachments/assets/a6093330-8340-477e-b0a1-bde120b8b33d)
